components/org.wso2.carbon.identity.oauth2.grant.organizationswitch/src/main/java/org/wso2/carbon/identity/oauth2/grant/organizationswitch/OrganizationSwitchGrant.java

```diff
@@ -183,7 +188,17 @@ private void checkOrganizationIsAllowedToSwitch(String currentOrgId, String swit
                 throw new IdentityOAuth2ClientException("Organization switch is only allowed for the organizations " +
                         "in the same branch.");
             }
+            if (!OrganizationSwitchGrantConstants.CONSOLE_APP_NAME.equals(appName) &&
+                    getOrgApplicationManager().resolveSharedApplication(appName, currentOrgId, switchOrgId) == null) {
+                throw new IdentityOAuth2ClientException("Organization switching is not allowed for organizations " +
+                        "that have not shared the application");
+            }
         } catch (OrganizationManagementException e) {
+            if (e.getErrorCode().equals(
+                    OrganizationManagementConstants.ErrorMessages.ERROR_CODE_APPLICATION_NOT_SHARED.getCode())) {
+                throw new IdentityOAuth2ClientException("Organization switching is not allowed for organizations " +
+                        "that have not shared the application");
+            }
```

Comment:
We could have taken out the entire new logic with try-catch block out side the existing try catch block which is added for organization depth checking. 
That will simplify the two validations which is about to happen.